Add security best practices page. (#1820)
graphql/graphql.github.io: src/pages/learn/security.mdx

          name
        }
      }
    }
  }
}
```

Read more about the connection-based model for paginating fields on the [Pagination page](/learn/pagination/).

### Depth limiting

One of GraphQL's strengths is that clients can write expressive operations that reflect the relationships between the data exposed in the API. However, some queries may become cyclical when the fields in the selection set are deeply nested: 

```graphql
query {
  hero {
    name
    friends {
      name
      friends {
        name
        friends {
          name
          friends {
            name
          }
        }
      }
    }
  }
}
```

Even when [the N+1 problem](/learn/performance/#the-n1-problem) has been remediated through batched requests to underlying data sources, overly nested fields may still place excessive load on server resources and impact API performance. 

For this reason, it's a good idea to limit the maximum depth of fields that a single operation can have. Many GraphQL implementations expose configuration options that allow you to specify a maximum depth for a GraphQL document and return an error to the client if a request exceeds this limit before execution begins.

> For cases where a client may have a legitimate use case for a deeply nested query and it's impractical to set a blanket limit on all queries, you may instead opt for applying depth limits specifically to list fields instead, or using [rate limiting](/#rate-limiting) instead.

Reviewer: List limits should be used in general, they're much more vital than simple depth limits alone. This is how GraphQL.js protects against malicious introspection queries for example: https://github.com/graphql/graphql-js/blob/4ef0428602fe2e16fd075e1e3e51b1feca9aad38/src/validation/rules/MaxIntrospectionDepthRule.ts#L9

I've separated this out as its own recommendation, and added query complexity analysis as another option for dealing with this. (A heading will need to be added below for this link to be valid.)

```suggestion
Since nesting list fields can result in exponential increases to the amount of data returned, it's recommended to apply a separate smaller limit to how deeply lists can be nested.

For cases where a client may have a legitimate use case for a deeply nested query and it's impractical to set a low blanket limit on all queries, you may need to rely on techniques such as [rate limiting](/#rate-limiting) or [query complexity analysis](#query-complexity-analysis).
```

### Breadth and batch limiting

In addition to limiting operation depth, there should also be guardrails in place to limit the number of top-level fields and field aliases included in a single operation.

Consider what would happen during the execution of the following query operation:

```graphql
query {
  hero1: hero(episode: NEWHOPE) {
    name
  }
  hero2: hero(episode: EMPIRE) {
    name
  }
  ### ...
  hero100: hero(episode: JEDI) {
    name
  }

> }

Reviewer: `DataLoader` would likely absorb this and turn it into a single request to resolve the data. This is a valid concern, but the example is perhaps not the best illustration of it. Maybe something that couldn't easily be data-loaded would be better - e.g.

```suggestion
query {
  viewer {
    friends1: friends(limit: 1) { name }
    friends2: friends(limit: 2) { name }
    friends3: friends(limit: 3) { name }
    # ...
    friends100: friends(limit: 100) { name }
  }
}
```

(If accepted, this may need adjusting to fit the other example's schema, and the paragraph below will need updating too.)

Similarly, a client may send a GraphQL document with many batched operations in a request:

```graphql
query NewHopeHero {
  hero(episode: NEWHOPE) {
    name
  }
}
query EmpireHero {
  hero(episode: EMPIRE) {
    name
  }
}
### ...
query JediHero {
  hero(episode: JEDI) {
    name
  }
}
```

Depending on the client implementation, query batching can be a useful strategy for limiting the number of round trips to a server to fetch all of the required data to render a user interface. However, there should be an upper limit on the total number of queries allowed in a single batch.

As with depth limiting, a GraphQL implementation may have configuration options to restrict operation breadth, field alias usage, and batching.

### Rate limiting

Depth, breadth, and batch limiting help prevent broad categories of malicious operations such as cyclic queries and batching attacks, but they don't provide a way to declare that a particular field is computationally expensive to resolve. So for more advanced demand control requirements, you may wish to implement rate limiting. 

> Rate limiting may take place in different layers of an application, for example, in the network layer or the business logic layer. Because GraphQL allows clients to specify exactly what data they need in their queries, a server may not be able to know in advance if a request includes fields that will place a higher load on its resources during execution. As a result, applying useful rate limits for a GraphQL API typically requires a different approach than simply keeping track of the total number of incoming requests over a time period in the network layer.

Reviewer: ```suggestion
Rate limiting may take place in different layers of an application, for example, in the network layer or the business logic layer. Because GraphQL allows clients to specify exactly what data they need in their queries, a server may not be able to know in advance if a request includes fields that will place a higher load on its resources during execution. As a result, applying useful rate limits for a GraphQL API typically requires a different approach than simply keeping track of the total number of incoming requests over a time period in the network layer; therefore applying rate limits within the business logic layer is generally recommended.
```

> _Query complexity analysis_ is one method that can be used to rate limit GraphQL requests by applying weights to the types and fields in a schema and then analyzing an incoming document to determine if the combination of fields included in its selection set exceeds a maximum allowable cost per request. If the request proceeds, the total cost of the request can be deducted from the overall query budget allocated for a specific period.

Reviewer: I think Query Complexity Analysis warrants its own section since you can use it as a simple complexity cap without needing to tie it into rate limiting. (Rate limiting has a higher infrastructure overhead, so it's worth pointing this out as an independently valuable technique to people who may not want to run that infrastructure yet.)

Since I've added the header, I've also reordered the sentence to avoid repetition, and tried to highlight that factoring this into rate limits is optional, but valuable.

```suggestion
### Query complexity analysis

By applying weights to the types and fields in a schema, you can estimate the cost of incoming requests using the technique known as _query complexity analysis_. If the combination of fields included in a request exceeds a maximum allowable cost per request, you may choose to reject the request outright. The estimated cost can also be factored into rate limiting: if the request proceeds, the total cost of the request can be deducted from the overall query budget allocated to a client for a specific time period.
```

> While the GraphQL specification doesn't provide any guidelines on implementing query complexity analysis or rate limits for an API, there is [a community-maintained draft specification]((https://ibm.github.io/graphql-specs/cost-spec.html)) for implementing custom type system directives that support these calculations.

Reviewer: Typo?

```suggestion
While the GraphQL specification doesn't provide any guidelines on implementing query complexity analysis or rate limits for an API, there is [a community-maintained draft specification](https://ibm.github.io/graphql-specs/cost-spec.html) for implementing custom type system directives that support these calculations.
```

### Trusted documents

For GraphQL APIs that only serve first-party clients, using _trusted documents_ will allow you to create an allowlist of operations that can be executed against a schema.

> As a build step during development, clients may submit their GraphQL documents to the server's allowlist to be stored using the hashed document string as an ID. At runtime, the client can send the hashed document ID instead of the full GraphQL document, and the server will only execute the request for known document IDs.

Reviewer: Trusted documents don't have to use a hash, so I've edited this for precision:

```suggestion
As a build step during development clients may submit their GraphQL documents to the server's allowlist, where each document is stored with a unique ID—usually its hash. At runtime, the client can send a document ID instead of the full GraphQL document, and the server will only execute requests for known document IDs.
```

> The GraphQL specification doesn't require the use of trusted documents or indicate how they should be implemented, but there are an increasing number of GraphQL implementations that support them.

Reviewer: Since persisted documents are a transport level concern, they're out of scope of the GraphQL spec; I've updated this text to link to the persisted documents GraphQL-over-HTTP RFC which is already fairly widely implemented, explain the reason for the "trusted" in trusted documents, and outline the difference between them and persisted documents/queries. What do you think?

```suggestion
Trusted documents are simply _persisted documents_ that are deemed to not be malicious, usually because they were authored by your developers and approved through code review. Efforts are underway to [standardize _persisted documents_ as part of the GraphQL-over-HTTP specification](https://github.com/graphql/graphql-over-http/pull/264)—an increasing number of GraphQL clients and servers already support them (sometimes under their legacy misnomer: _persisted queries_).
```

Author: I think this helps provide better clarification.

## Schema considerations

### Validating and sanitizing argument values

GraphQL is strongly typed, and the [validation stage](/learn/validation/) of a request will check a parsed document against the type system to determine its validity. However, for certain cases you may find that the built-in Scalar types aren't enough to make sure that the argument value a client provides is appropriate to execute a field resolver safely.

Consider the following `ReviewInput` type that is used with the `createReview` mutation: 

```graphql
input ReviewInput {
  stars: Int!
  commentary: String
}

type Mutation {

>   createReview(episode: Episode, review: ReviewInput!): Review

Reviewer: If this isn't part of a larger example I'm not aware of, removing the `Episode` type may simplify this for the reader:

```suggestion
  createReview(episodeId: ID!, review: ReviewInput!): Review
```

Author: This is how it's defined in the example schema:

https://github.com/graphql/graphql.github.io/blob/1dd7f814e0223c9a2b4950cf0e2f8daaae97b92b/src/components/marked/swapi-schema.tsx#L32-L34

Reviewer: Oh `Episode` is an enum! I thought it was an object type! This is fine then.

}
```

Depending on what kind of client application will submit and display the review data, it may be necessary to sanitize the string provided for the `commentary` field in case a user submits unsafe HTML in it.

The sanitization step should take place in the business logic layer that is called during execution of the `createReview` resolver function. To surface this to frontend developers, it could also be expressed as a part of the schema using a custom [Scalar type](/learn/schema/#scalar-types).

### Introspection

[Introspection](/learn/introspection/) is a powerful feature of GraphQL that allows you to query a GraphQL API for information about its schema. However, GraphQL APIs that only serve first-party clients may forbid introspection queries in non-development environments.

> Note that disabling introspection is a form of "security through obscurity" and will not be sufficient to protect a GraphQL API by itself, but it may be used as a part of a broader security strategy to limit the discoverability of API information from potential attackers. For more comprehensive protection of sensitive schema details and user data, trusted documents and authorization should be used.

Reviewer: Disabling introspection is good to protect against introspection attacks that attempt to achieve denial of service; hopefully this edit clarifies the use case we're talking about here:

```suggestion
Note that disabling introspection can be used as a form of "security through obscurity," but will not be sufficient to entirely obscure a GraphQL API by itself. It may be used as a part of a broader security strategy to limit the discoverability of API information from potential attackers. For more comprehensive protection of sensitive schema details and user data, trusted documents and authorization should be used.
```

### Error messages

On the [Response page](/learn/response/), we saw that a GraphQL implementation may provide developers with useful information about validation errors in a request, along with suggestions for how these errors may be addressed. For example:

```graphql
# { "graphiql": true }
query {
  starship(id: 3000) {
    width
  }
}
```

> These hints can be helpful when debugging client-side errors, but they may provide more information about a schema in a production environment than we would like to reveal. Hiding detailed error information in GraphQL responses outside of development environments is important because, even with introspection disabled, an attacker could ultimately infer the shape of an entire schema by running numerous operations with incorrect field names.

Reviewer: In addition to validation errors ("request errors" in GraphQL spec terms), it would be wise to also talk about runtime errors ("field errors" in GraphQL spec terms) and error masking.

Author: Resolved in a new commit.

Since nesting list fields can result in exponential increases to the amount of data returned, it's recommended to apply a separate smaller limit to how deeply lists can be nested.

For cases where a client may have a legitimate use case for a deeply nested query and it's impractical to set a low blanket limit on all queries, you may need to rely on techniques such as [rate limiting](/#rate-limiting) or [query complexity analysis](#query-complexity-analysis).

### Breadth and batch limiting

In addition to limiting operation depth, there should also be guardrails in place to limit the number of top-level fields and field aliases included in a single operation.

Consider what would happen during the execution of the following query operation:

```graphql
query {
  viewer {
    friends1: friends(limit: 1) { name }
    friends2: friends(limit: 2) { name }
    friends3: friends(limit: 3) { name }
    # ...
    friends100: friends(limit: 100) { name }
  }
}
```

Even though the overall depth of this query is shallow, the underlying data source for the API will still have to handle a large number of requests to resolve data for the aliased `hero` field.

> Similarly, a client may send a GraphQL document with many batched operations in a request:

Reviewer 2: Is `batching` standard to GraphQL? I think it's not in the official specification: https://spec.graphql.org/October2021/ 

It's only on *some* GraphQL clients.

Reviewer: The official specification doesn't even cover HTTP. We're in the process of specifying batching as part of the GraphQL-over-HTTP working group:

- https://github.com/graphql/graphql-over-http/pull/307
- https://github.com/graphql/graphql-over-http/pull/308

```graphql
query NewHopeHero {
  hero(episode: NEWHOPE) {
    name
  }
}
query EmpireHero {
  hero(episode: EMPIRE) {
    name
  }
}
### ...
query JediHero {
  hero(episode: JEDI) {
    name
  }
}
```

Depending on the client implementation, query batching can be a useful strategy for limiting the number of round trips to a server to fetch all of the required data to render a user interface. However, there should be an upper limit on the total number of queries allowed in a single batch.

As with depth limiting, a GraphQL implementation may have configuration options to restrict operation breadth, field alias usage, and batching.

> 

Reviewer 2: Would be nice to add paragraph about recursive fragments :D

![CleanShot 2024-11-18 at 17 34 09@2x](https://github.com/user-attachments/assets/5a5bc984-1532-44e4-92a7-6e772f45e7de)

Reviewer: A GraphQL API cannot have recursive fragments, as stated by the GraphQL specification:

https://spec.graphql.org/draft/#sec-Fragment-Spreads-Must-Not-Form-Cycles

Reviewer 2: Interesting thanks!

Note: seems like it’s not always implemented properly — a CVE we found with our team.

https://github.com/graphql-rust/juniper/security/advisories/GHSA-4rx6-g5vg-5f3j

Reviewer: That definitely seems like a bug in that implementation, missing one of the core validation rules is quite a major diversion from the spec - I would expect that to be opt-in.

It would be really good to get the https://github.com/graphql-cats/graphql-cats project (or one like it) up and running again so it can be checked if an implementation is actually spec-compliant. If anyone was interested in taking this on, I'm fairly sure it would likely get approval from the [community grant program](https://graphql.org/community/foundation/community-grant/).

### Rate limiting

Depth, breadth, and batch limiting help prevent broad categories of malicious operations such as cyclic queries and batching attacks, but they don't provide a way to declare that a particular field is computationally expensive to resolve. So for more advanced demand control requirements, you may wish to implement rate limiting. 

Rate limiting may take place in different layers of an application, for example, in the network layer or the business logic layer. Because GraphQL allows clients to specify exactly what data they need in their queries, a server may not be able to know in advance if a request includes fields that will place a higher load on its resources during execution. As a result, applying useful rate limits for a GraphQL API typically requires a different approach than simply keeping track of the total number of incoming requests over a time period in the network layer; therefore applying rate limits within the business logic layer is generally recommended.

> 

Reviewer 2: The perfect common example here is trying to by pass authentication server rate limiting using aliasing or batching. 

![CleanShot 2024-11-18 at 17 40 07@2x](https://github.com/user-attachments/assets/274e5d26-c6fb-454a-9abc-9bccaf8e17c1)

### Query complexity analysis

By applying weights to the types and fields in a schema, you can estimate the cost of incoming requests using the technique known as _query complexity analysis_. If the combination of fields included in a request exceeds a maximum allowable cost per request, you may choose to reject the request outright. The estimated cost can also be factored into rate limiting: if the request proceeds, the total cost of the request can be deducted from the overall query budget allocated to a client for a specific time period.

While the GraphQL specification doesn't provide any guidelines on implementing query complexity analysis or rate limits for an API, there is [a community-maintained draft specification](https://ibm.github.io/graphql-specs/cost-spec.html) for implementing custom type system directives that support these calculations.

### Trusted documents

> For GraphQL APIs that only serve first-party clients, using _trusted documents_ will allow you to create an allowlist of operations that can be executed against a schema.

Reviewer 3: I would maybe put Trusted Documents at the start?
Mostly because of how important they are and also because its one of the biggest misconceptions around GraphQL and security?

I would add maybe something along the lines of - 
GraphQL is mostly use as a developer tool - meaning you can use queries as you develop your app and then lock your queries in production.
That means that you can lock in advance only the queries that your team needs.
Only in case you need public APIs, and can't use Trusted Documents, then you are leaving your API exposed to all these attacks and can use tools to defend against all the following attacks..

Author: Addressed in the latest commit.

As a build step during development clients may submit their GraphQL documents to the server's allowlist, where each document is stored with a unique ID—usually its hash. At runtime, the client can send a document ID instead of the full GraphQL document, and the server will only execute requests for known document IDs.

> Trusted documents are simply _persisted documents_ that are deemed to not be malicious, usually because they were authored by your developers and approved through code review. Efforts are underway to [standardize _persisted documents_ as part of the GraphQL-over-HTTP specification](https://github.com/graphql/graphql-over-http/pull/264)—an increasing number of GraphQL clients and servers already support them (sometimes under their legacy misnomer: _persisted queries_).

Reviewer 4: people also call these _persisted operations_

Reviewer: Also "stored queries" and "stored operations" (and maybe "stored documents", but I don't recall that?) have been used. I think that's too much detail to include in these docs though; "persisted queries" is sufficiently pervasive to warrant mentioning but the others can be glossed over.

The sanitization step should take place in the business logic layer that is called during execution of the `createReview` resolver function. To surface this to frontend developers, it could also be expressed as a part of the schema using a custom [Scalar type](/learn/schema/#scalar-types).

### Introspection

[Introspection](/learn/introspection/) is a powerful feature of GraphQL that allows you to query a GraphQL API for information about its schema. However, GraphQL APIs that only serve first-party clients may forbid introspection queries in non-development environments.

Note that disabling introspection can be used as a form of "security through obscurity," but will not be sufficient to entirely obscure a GraphQL API by itself. It may be used as a part of a broader security strategy to limit the discoverability of API information from potential attackers. For more comprehensive protection of sensitive schema details and user data, trusted documents and authorization should be used.

> 

Reviewer 2: Since you already added considerations about `batching` which I think is not standard in GraphQL, you can also talk about `Field Suggestion` which exits in *some* GraphQL clients.

![CleanShot 2024-11-18 at 17 42 31@2x](https://github.com/user-attachments/assets/341bcd25-9250-478c-8309-ec277399ef47)

Reviewer: This is one of many many ways of extracting schema information despite introspection being disabled, I worry if we list this as an example then people will think they just need to solve that too and then they're good - they're not. Ultimately: GraphQL schemas should not be treated as secrets/if your schema is secret then it should not accept arbitrary operations (i.e. you should use trusted documents).

If you're interested, a little more information on this topic can be found in my GraphQLConf 2024 talk; here's a timestamped link to the relevant part: https://youtu.be/Ytt1_ZIlYdg?t=1267

Reviewer 2: "Looking over the shoulder" is a fun one ;)
But that's totally true!

And Nice talk and thanks for the various mentions 🙏
